File: 200 Number of Islands/number_of_islands.c
void dfs(char** grid, int gridRowSize, int gridColSize, int row, int col) {
    if (row < 0 || row >= gridRowSize || col < 0 || col >= gridColSize)
        return;
    if (grid[row][col] == '1') {
        grid[row][col] = '0';
        dfs(grid, gridRowSize, gridColSize, row - 1, col);
        dfs(grid, gridRowSize, gridColSize, row, col - 1);
        dfs(grid, gridRowSize, gridColSize, row + 1, col);
        dfs(grid, gridRowSize, gridColSize, row, col + 1);
    }
}

int numIslands(char** grid, int gridRowSize, int gridColSize) {
    if (gridRowSize == 0) return 0;
    int i, j, count = 0;
    for (i = 0; i < gridRowSize; i++) {
        for (int j = 0; j < gridColSize; j++) {
            if (grid[i][j] == '1') {
                count++;
                dfs(grid, gridRowSize, gridColSize, i, j);
            }
        }
    }
    return count;
} /* Author: XC */

File: 200 Number of Islands/number_of_islands.cs
public class Solution {
    private int count, rows, cols;
    private int[] dx = {-1, 1, 0, 0};
    private int[] dy = {0, 0, -1, 1};
    
    public int NumIslands(char[,] grid) {
        count = 0;
        rows = grid.GetLength(0);
        cols = grid.GetLength(1);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i, j] == '1') count++;
                dfs(grid, i, j);
            }
        }
        return count;
    }
    
    private void dfs(char[,] grid, int x, int y) {
        if (x < 0 || x >= rows || y < 0 || y >= cols) return;
        if (grid[x, y] == '2' || grid[x, y] == '0') return;
        char ch = grid[x, y];
        grid[x, y] = '2';
        for (int i = 0; i < 4; i++) {
            dfs(grid, x + dx[i], y + dy[i]);
        }
    }
} // Author: XC
